### DISCUSSION TITLE
OSD joystick not working

### DISCUSSION THREAD

**Discussion by neistow** [iNavFlight/inav#10513] [Support] [answered]
While in the field, I'm trying to set up an OSD joystick to tweak camera settings. My camera is Runcam Phoenix Pro and has menu wire soldered to the CC pad on the FC (Speedy Bee F405 V3). From the [docs](https://github.com/iNavFlight/inav/blob/master/docs/OSD%20Joystick.md), I don't quite understand if I can use an OSD joystick while the menu wire is soldered to the CC pad. 

What I've done so far, according to the guide:
- `set led_pin_pwm_mode=shared_high` 
- `set osd_joystick_enabled=on`
- Enabled Multi-color RGB LED Strip support in the configuration tab

But when trying to use stick commands to enter the camera menu, nothing happens. Moreover, when `osd_joystick_enabled` is ON, I can't use stick commands to get into the regular OSD menu.

Can somebody explain me how this intended to work?

**Comment by MrD-RC** [comment]
Camera control is not supported in INAV AFAIK.

**Comment by sensei-hacker** [comment]
Hmm @RomanLut has made a number of updates to the camera control code src/main/io/osd_joystick.c and documentation docs/OSD Joystick.md

**Comment by RomanLut** [comment, marked as answer]
> While in the field, I'm trying to set up an OSD joystick to tweak camera settings. My camera is Runcam Phoenix Pro and has menu wire 

1) Which version of INAV are you using?  I will double check  'when osd_joystick_enabled is ON, I can't use stick commands to get into the regular OSD menu.' - it should not happend. Also it might be the case that you entered OSD joystick mode (Throttle:CENTER Yaw:RIGHT). You have to leave OSD Joystick mode ( Throttle:CENTER Yaw:LEFT ) to be able to enter OSD CMS menu.

2) It will not work with menu wire connected directly to LED pin. You have to build RC filter as specified in the docs: 
![image](https://github.com/user-attachments/assets/9aaae105-0947-4a75-a229-436ee8dfcf42)
Where ```FC_LED_PIN``` is connected to Flight conttrller ```LED``` pin, and ```OSD_JOYSTICK``` to ```camera menu pin```. I would test with '''WS2811_LED_STRIP``` wire disconnected first.

3) You might need to to adjust ```osd_joystick_xxxx``` values as specified in the docs, but ```DOWN``` key should work in any case.

**Reply by neistow** [reply]
Thanks for the quick reply. 

> Also it might be the case that you entered OSD joystick mode (Throttle:CENTER Yaw:RIGHT). You have to leave OSD Joystick mode ( Throttle:CENTER Yaw:LEFT ) to be able to enter OSD CMS menu

I've done some testing and this was indeed true. I did enter the OSD joystick mode, but I don't see camera menu on my video feed, like I did when entered camera menu with physical control joystick connected. After exiting joystick mode I was able to get into OSD CMS menu.

> You might need to to adjust osd_joystick_xxxx values as specified in the docs, but DOWN key should work in any case.

If I understand correctly, then it's osd_joystick_xxxx values I need to tweak in order for the menu to appear, because camera doesn't see proper response from the FC? If so, do I really need RC filter (I don't have one) and doing all that steps, or I can let's say manually guess numbers or at least deduce, because I had camera OSD menu working when plugged joystick from caddx ratel.

Does betaflight `camera_control_button_resistance = 450,270,150,68,0` setting correlate with inav `osd_joystick_xxxx` setting somehow? I really wonder if betaflight default settings will work for my camera, but at this point I don't really want to flash it

**Reply by RomanLut** [reply]
It won't work without an RC filter. The FC cannot directly generate the voltage levels required by the camera; instead, it produces a PWM signal, which the RC filter converts into voltage. By installing an RC filter, you likely won't need to adjust the PWM values further.

Betaflight settings do not corellate with INAV settings.

**Reply by neistow** [reply]
What about the option to guess the values? I can see some logical order in values. What if I tweak each value adjusting by 2-5-10, saving, and trying out? Is there some range for the pwm values that works: like `osd_joystick_right` is 28 by default, assuming this value works, will 27 and 29 also work?

What I want to try to do is to find at least one pwm value, presumably, `osd_joystick_right`, since setting this value correctly will at least show me the camera OSD menu am I correct?

**Reply by RomanLut** [reply]
Yes you can guess values by tring all 0,5,10...100 but you will complete task much faster if you use multimeter as described in the docs. Yes close values will work. You can guess minimum and maximum working values and use the middle.

**Reply by neistow** [reply]
At this point, I have tried setting all 0-100 values one by one for `osd_joystick_enter`, but neither of those values didn't get me the camera menu on my OSD. I made sure I had entered joystick emulation mode (by trying to enter osd cms menu). Maybe, I have something other wrongly configured, but I made sure `led_pin_pwm_mode=shared_high` and multicolor rgb strip support was turned on.

**Reply by RomanLut** [reply]
So, did  you installed RC Filter?
What INAV version are you using?

**Reply by neistow** [reply]
No, I didn't install RC Filter, because I don't have one. Is it necessary to have RC Filter in order feature to work? Just having menu wire soldered to CC pad and correct osd_joystick values is not enough? My wiring is according to this diagram:

![image](https://github.com/user-attachments/assets/049799d3-1280-4b7d-86a1-4dd442d8ab2c)

I'm using INAV 7.1.2.

**Reply by RomanLut** [reply]
It will not work with menu wire connected directly to LED pin. You have to build RC filter as specified in the docs.

**Reply by neistow** [reply]
so, CC pad is the same as D4 pin?

**Reply by RomanLut** [reply]
No CC is UART? TX and is used exclusively for SmartAudio/Trump.

For OSD joystick emulation, an RC filter must be connected to the LED pin. The RC filter consists of a 470-ohm resistor and a 10µF capacitor. These components can be easily salvaged from old electronic devices.

**Reply by neistow** [reply]
> No CC is UART? 

Afaik - no? Betaflight guide to [fpv camera control](https://betaflight.com/docs/wiki/guides/current/fpv-camera-control-joystick-emulation#how-do-i-select-a-suitable-pin) suggests using CC pad, so I assumed it would work in INAV too

> For OSD joystick emulation, an RC filter must be connected to the LED pin. The RC filter consists of a 470-ohm resistor and a 10µF capacitor. These components can be easily salvaged from old electronic devices.

That's an option, but at this point, for me, it's easier to plug in a harness for the OSD joystick and leave it in place, maybe zip-tie it, so it doesn't dangle around. Making an RC filter will take too much effort, which is not what I hoped for.

~~Could perhaps some approach similar to betaflight be implemented in INAV?~~ I read about resource remapping and found the video why it couldn't be implemented in inav currently


Thanks for the help, this discussion can be closed.
